File: cloud-admin/selective_repeat.py
"""
Selective Repeat ARQ — Cloud side (FR-MQTT2).

Tracks received chunks via bitmap, sends ACK for each received chunk,
and NACKs gaps within the sliding window.
"""

import time


class CloudSelectiveRepeat:
    def __init__(self, window_size=8, timeout_sec=5.0):
        self.window_size = window_size
        self.timeout_sec = timeout_sec
        self.expected_base = 0
        self.received_bitmap = None  # initialized when transfer starts
        self.total_chunks = 0
        self.last_received_time = {}  # seq -> timestamp
        self.ack_callback = None  # set by mqtt_admin to publish ACK/NACK

    def start_session(self, total_chunks):
        """Initialize state for a new file transfer session."""
        self.total_chunks = total_chunks
        bitmap_size = (total_chunks + 7) // 8
        self.received_bitmap = bytearray(bitmap_size)
        self.expected_base = 0
        self.last_received_time = {}

    def on_chunk_received(self, seq_num):
        """Mark chunk as received, send ACK, detect gaps and NACK."""
        # Set bit in bitmap
        byte_idx = seq_num // 8
        bit_idx = seq_num % 8
        self.received_bitmap[byte_idx] |= (1 << bit_idx)
        self.last_received_time[seq_num] = time.time()

        # Always ACK received chunk
        if self.ack_callback:
            self.ack_callback("ACK", seq_num)

        # Advance expected_base
        while self.expected_base < self.total_chunks:
            b = self.expected_base // 8
            bit = self.expected_base % 8
            if self.received_bitmap[b] & (1 << bit):
                self.expected_base += 1
            else:
                break

        # Check for gaps within window — NACK missing chunks
        for i in range(self.expected_base, min(self.expected_base + self.window_size, self.total_chunks)):
            b = i // 8
            bit = i % 8
            if not (self.received_bitmap[b] & (1 << bit)):
                if self.ack_callback:
                    self.ack_callback("NACK", i)

    def check_timeouts(self):
        """Check for chunks not received within timeout, send NACK."""
        now = time.time()
        for seq in range(self.expected_base, min(self.expected_base + self.window_size, self.total_chunks)):
            b = seq // 8
            bit = seq % 8
            if not (self.received_bitmap[b] & (1 << bit)):
                # Not received — NACK if we haven't recently
                last = self.last_received_time.get(seq)
                if last is None or (now - last) > self.timeout_sec:
                    if self.ack_callback:
                        self.ack_callback("NACK", seq)

    def is_complete(self):
        """Return True if all chunks have been received."""
        return self.expected_base >= self.total_chunks

    def progress(self):
        """Return transfer progress as a percentage."""
        if self.total_chunks == 0:
            return 0.0
        received = sum(bin(b).count('1') for b in self.received_bitmap) if self.received_bitmap else 0
        return received / self.total_chunks * 100.0

File: cloud-admin-go/selective_repeat.go
package main

import "time"

// CloudSelectiveRepeat implements a receiver-side selective repeat protocol
// for the cloud admin. It tracks received chunks via a bitmap and issues
// NACKs for missing chunks within the receive window.
//
// Bug fix vs Python version: NACKs are decoupled from ACKs. OnChunkReceived
// only ACKs; CheckTimeouts is the sole NACK source, with a per-sequence
// cooldown to prevent NACK flooding.
type CloudSelectiveRepeat struct {
	windowSize   int
	timeoutSec   float64
	expectedBase int
	bitmap       []byte
	totalChunks  int
	lastNACKTime map[int]float64 // per-seq NACK cooldown
	ackCallback  func(msgType string, seq int)
	NACKCount    int // total NACKs sent during this session
}

// NewCloudSelectiveRepeat creates a new SR instance.
func NewCloudSelectiveRepeat(windowSize int, timeoutSec float64) *CloudSelectiveRepeat {
	return &CloudSelectiveRepeat{
		windowSize:   windowSize,
		timeoutSec:   timeoutSec,
		lastNACKTime: make(map[int]float64),
	}
}

// StartSession initialises tracking state for a transfer with totalChunks fragments.
func (sr *CloudSelectiveRepeat) StartSession(totalChunks int) {
	sr.totalChunks = totalChunks
	bitmapSize := (totalChunks + 7) / 8
	sr.bitmap = make([]byte, bitmapSize)
	sr.expectedBase = 0
	sr.lastNACKTime = make(map[int]float64)
	sr.NACKCount = 0
}

// isReceived checks whether the bit for seq is set.
func (sr *CloudSelectiveRepeat) isReceived(seq int) bool {
	if seq < 0 || seq >= sr.totalChunks {
		return false
	}
	return sr.bitmap[seq/8]&(1<<uint(seq%8)) != 0
}

// markReceived sets the bit for seq.
func (sr *CloudSelectiveRepeat) markReceived(seq int) {
	if seq >= 0 && seq < sr.totalChunks {
		sr.bitmap[seq/8] |= 1 << uint(seq%8)
	}
}

// OnChunkReceived marks seq as received, advances the base pointer, and
// sends an ACK via the callback. It does NOT send NACKs — that is handled
// exclusively by CheckTimeouts.
func (sr *CloudSelectiveRepeat) OnChunkReceived(seq int) {
	if seq < 0 || seq >= sr.totalChunks {
		return
	}
	sr.markReceived(seq)

	// Advance base past contiguous received range.
	for sr.expectedBase < sr.totalChunks && sr.isReceived(sr.expectedBase) {
		sr.expectedBase++
	}

	// ACK this chunk.
	if sr.ackCallback != nil {
		sr.ackCallback("ACK", seq)
	}
}

// CheckTimeouts sends NACKs for missing chunks in the current window, but
// only if at least 2 seconds have elapsed since the last NACK for that seq.
func (sr *CloudSelectiveRepeat) CheckTimeouts() {
	if sr.bitmap == nil {
		return
	}
	now := float64(time.Now().UnixMilli()) / 1000.0
	end := sr.expectedBase + sr.windowSize
	if end > sr.totalChunks {
		end = sr.totalChunks
	}
	const nackCooldown = 2.0 // seconds

	for seq := sr.expectedBase; seq < end; seq++ {
		if sr.isReceived(seq) {
			continue
		}
		lastTime, exists := sr.lastNACKTime[seq]
		if exists && (now-lastTime) < nackCooldown {
			continue
		}
		sr.lastNACKTime[seq] = now
		sr.NACKCount++
		if sr.ackCallback != nil {
			sr.ackCallback("NACK", seq)
		}
	}
}

// IsComplete returns true when every chunk has been received.
func (sr *CloudSelectiveRepeat) IsComplete() bool {
	return sr.expectedBase >= sr.totalChunks
}

// Progress returns the fraction of chunks received (0.0 – 1.0).
func (sr *CloudSelectiveRepeat) Progress() float64 {
	if sr.totalChunks == 0 {
		return 0
	}
	count := 0
	for seq := 0; seq < sr.totalChunks; seq++ {
		if sr.isReceived(seq) {
			count++
		}
	}
	return float64(count) / float64(sr.totalChunks)
}
